File: haskell/hitori.hs
import Data.Array
import Data.List

type Board = Array (Int, Int) Int

boardH :: Int
boardH = 8

boardW :: Int
boardW = 8

-- Move all array's elemnts in a list of lists
arrayToList :: Board -> [[Int]]
arrayToList a =
    [[a ! (r,c) | c <- [1..boardW]] | r <- [1..boardH]] 
---------------------------------------------------------------------------------------------------------------------------------------------------
-- 1 Rule: no adjacent shadowed cells

-- Check ther're no adjacent shadowed cells
checkNoAdjB :: [Int] -> Bool
checkNoAdjB [] = True
checkNoAdjB [x] = True
checkNoAdjB (x:y:xs) 
    | x == y = False
    | otherwise = checkNoAdjB (y:xs)

checkNABinR :: [[Int]] -> Bool
checkNABinR [] = True
checkNABinR grid = all checkNoAdjB grid 

checkNABinC :: [[Int]] -> Bool
checkNABinC [] = True
checkNABinC grid = all checkNoAdjB (transpose grid) 

checkFirtsRule :: Board -> Bool
checkFirtsRule b = 
    let grid = arrayToList b
    in checkNABinC grid && checkNABinR grid 
---------------------------------------------------------------------------------------------------------------------------------------------------
-- 2 Rule: no duplicated elements in each row / cell

checkNoDuplicated :: [Int] -> Bool
checkNoDuplicated [] = True
checkNoDuplicated (x:xs) 
    | x /= 0 && elem x xs = False   -- There's a duplicated element
    | otherwise = checkNoDuplicated xs  -- Go forward checking the list

checkNDinR :: [[Int]] -> Bool
checkNDinR [] = True
checkNDinR grid = all checkNoDuplicated grid

checkNDinC :: [[Int]] -> Bool
checkNDinC [] = True
checkNDinC grid = all checkNoDuplicated (transpose grid)

checkSecondRule :: Board -> Bool
checkSecondRule b = 
    let grid = arrayToList b
    in checkNDinC grid && checkNDinR grid
---------------------------------------------------------------------------------------------------------------------------------------------------
-- 3 Rule: non-shadowed cells must be connected

-- Flooding to check if the unshadowed (/= 0) elements are connected
fill :: Board -> (Int, Int) -> Board
fill board (x,y) = floodFill board [(x,y)]
    where
        -- Flood fill with list of cells
        floodFill :: Board -> [(Int, Int)] -> Board
        floodFill b [] = b
        floodFill b (pos:rest)
            | not (inBounds pos) = floodFill b rest   -- Out board, ignore
            | b ! pos == 0 = floodFill b rest -- Already checked or shadowed cell, go forward
            | otherwise = 
                let b' = b // [(pos, 0)]    -- Mark the cell with 0
                    neighbors = directions pos
                in floodFill b' (rest ++ neighbors)
        
        -- Get adj cells
        directions :: (Int, Int) -> [(Int, Int)]
        directions (i,j) = [(i-1, j), (i+1, j), (i, j-1), (i, j+1)]

        -- Check position within the board
        inBounds :: (Int, Int) -> Bool
        inBounds (r, c) = 1 <= r && r <= boardH && 1 <= c && c <= boardW 

-- Find the first cell in the board /= 0
firstNonZero :: Board -> Maybe (Int, Int)
firstNonZero b = 
    case filter (\pos -> b ! pos /= 0) (indices b) of
        [] -> Nothing
        (x:_) -> Just x

checkRules :: Board -> Bool
checkRules b = 
    let
        firstRes = checkFirtsRule b
        secondRes = checkSecondRule b
        thirdRes = case firstNonZero b of
            Nothing -> False
            Just start -> do
                let r = fill b start
                if all (==0) (elems r) then True else False
    in firstRes && secondRes && thirdRes    

-- Main function
main :: IO ()
main = do
    -- Read the board from a file
    content <- readFile "hitori.txt"
    let rows = lines content
        boardList = concatMap (map read . words) rows :: [Int]
    
    -- Create the board as a 2D array
    let board = array ((1, 1), (boardH, boardW)) 
                      [((i, j), boardList !! ((i-1) * boardW + (j-1))) | i <- [1..boardH], j <- [1..boardW]]
    printBoard board

    case checkRules board of
        True -> putStrLn "Well done!"
        False -> putStrLn "The board is not valid"

  where
    -- Function to print the board
    printBoard b = mapM_ (\i -> putStrLn $ unwords [show (b ! (i, j)) | j <- [1..boardH]]) [1..boardW]

File: rust/hitori.rs
use std::collections::HashSet;
use std::fs;

const BOARD_H: usize = 8;
const BOARD_W: usize = 8;

type Board = Vec<Vec<i32>>;

// Read the board from .txt file
fn read_board(filename: &str) -> Board {
    let content = fs::read_to_string(filename).expect("Error reading the file");
    content
        .lines()
        .map(|line| {
            line.split_whitespace()
                .map(|x| x.parse::<i32>().expect("Error parsing cell content"))
                .collect()
        })
        .collect()
}

// Print the board
fn print_board(board: &Board) {
    for row in board {
        println!("{}",
            row.iter()
                .map(|x| x.to_string())
                .collect::<Vec<String>>()
                .join(" ")
        );
    }
}

// 1 Rule: no adjacent shadowed cells 
fn check_no_adj_shadowed(board: &Board) -> bool {
    for i in 0..BOARD_H {
        for j in 0..BOARD_W {
            if board[i][j] == 0 {
                // Check adj cells
                if (i > 0 && board[i-1][j] == 0)    // Up
                    || (i < BOARD_H-1 && board[i+1][j] == 0)    // Down
                    || (j > 0 && board[i][j-1] == 0)    // Left
                    || (j < BOARD_W-1 && board[i][j+1] == 0)    // Right
                {
                    return false;
                }
            }
        }
    }
    true
}

// 2 Rule: no duplicated elements in each row / cell
fn check_no_duplicates_in_vector(vec: &[i32]) -> bool {
    let mut seen = HashSet::new();
    for &val in vec {
        if val != 0 && !seen.insert(val) {  // insert(x) returns true if x is inserted, false if it's already present in HashSet
            return false;   // Duplicate value found
        }
    }
    true
}

fn check_no_duplicates(board: &Board) -> bool {
    // Check rows
    for r in board {
        if !check_no_duplicates_in_vector(r) {
            return false;
        }
    }

    // Check columns
    for c in 0..BOARD_W {
        let mut col = vec![];
        for r in 0..BOARD_H {
            col.push(board[r][c]);
        }
        if !check_no_duplicates_in_vector(&col) {
            return false;
        }
    }
    true
}

// 3 Rule: non-shadowed cells must be connected
fn check_connected(board: &Board) -> bool {

    // Find first un-shadowed cell
    let mut start = None;
    for (i, row) in board.iter().enumerate() {
        for (j, &val) in row.iter().enumerate() {
            if val != 0 {
                start = Some((i,j));  // Value found
                break;  // Stop inner cycle
            }
        }
        if start.is_some() {    // If a value has been found
            break;  // Stop outer cycle
        }
    }

    // If a valid value has been found (Some(x)) apply floodfill from it
    if let Some(start) = start {
        let mut visited = vec![vec![false; BOARD_W]; BOARD_H];  // Initialize to false an ancilla matrix
        flood_fill(board, start.0, start.1, &mut visited);

        // Check all un-shadowed cells have been visited
        for i in 0..BOARD_H {
            for j in 0..BOARD_W {
                if board[i][j] != 0 && ! visited[i][j] {
                    return false;
                }
            }
        }
        true
    } else {
        false   // There's no un-shadowed cells
    }
}

fn flood_fill(board: &Board, x: usize, y: usize, visited: &mut Vec<Vec<bool>>) {
    if x >= BOARD_H || y >= BOARD_W || visited[x][y] || board[x][y] == 0 {
        return;
    }
    visited[x][y] = true;

    // Check adj cells
    if x > 0 {
        flood_fill(board, x-1, y, visited);
    }
    if x < BOARD_H-1 {
        flood_fill(board, x+1, y, visited);
    }
    if y > 0 {
        flood_fill(board, x, y-1, visited);
    }
    if y < BOARD_W-1 {
        flood_fill(board, x, y+1, visited);
    }
}

// Check all rules
fn check_rules(board: &Board) -> bool {
    check_no_adj_shadowed(board) && check_no_duplicates(board) && check_connected(board)
}

// Main function
fn main() {
    let board = read_board("hitori.txt");
    println!("Board: ");
    print_board(&board);

    if check_rules(&board) {
        println!("Well done!");
    } else {
        println!("The board is not valid");
    }
}